### PR TITLE
test: add Unit tests to the ABM TF on GCE samples

### PR DESCRIPTION
**Pull request by Shabirmean**
#### I am still adding more tests. Opening the PR to make it easier to review

- Add unit tests for the 3 modules _(vm, external-ip, init)_ the main script under the `anthos-bm-gcp-terraform` sample
- Uses [gruntwork-io/terratest](https://github.com/gruntwork-io/terratest) and `golang json` for parsing the `terraform plan` and testing
- Tested values are the `variables` and the `planned_values` in the terraform plan

### CODE REVIEW

File: .github/workflows/ci_pr.yaml [hunk 1]
```diff
@@ -47,6 +47,9 @@ jobs:
       timeout-minutes: 20
       run: |
         golint ./... > golint.err 2>&1
+        if ! [ -s golint.err ]; then
```
Comment (askmeegs) [inline .github/workflows/ci_pr.yaml:50]: I think it's worth adding a ci-main workflow that runs on commits to the `main` branch, in addition to this PR workflow. the workflow can look identical.

File: .github/workflows/ci_pr.yaml [hunk 2]
```diff
@@ -47,6 +47,9 @@ jobs:
...
+          exit 0
+        fi
```
Comment (askmeegs) [inline .github/workflows/ci_pr.yaml:52]: Should the CI pipeline run the terraform unit tests?

**Comment by askmeegs** [comment]
How are you running these tests locally / with what variable values? I see 

```
Tested values are the variables and the planned_values in the terraform plan
```

But where are you getting these? Do you have test cases?

File: .github/workflows/ci_pr.yaml (hunk 1 above)
Comment (Shabirmean) [reply .github/workflows/ci_pr.yaml]: Yup. That's true. Ive added it in https://github.com/GoogleCloudPlatform/anthos-samples/pull/26/commits/4cdd344e7f7e9e387c38fc83d70ba0e974e6890a

File: .github/workflows/ci_pr.yaml (hunk 2 above)
Comment (Shabirmean) [reply .github/workflows/ci_pr.yaml]: Yes, sorry it was not committed when you reviewed it. A job for unit tests have been added here: https://github.com/GoogleCloudPlatform/anthos-samples/pull/26/files#diff-4becefff464f0184f0b2863fc9b24cf738be50aa1014a3c53483476e8027fc90R71-R93

I will add a README about running the unit tests for anyone using the repository

Comment (Shabirmean) [reply .github/workflows/ci_pr.yaml]: I've updated the contributing section in the [main README here](https://github.com/GoogleCloudPlatform/anthos-samples/tree/unit-tests)

Also added a section [specific to this sample here](https://github.com/GoogleCloudPlatform/anthos-samples/tree/unit-tests)

The expectation is have access atleast on Google Cloud Project to run the test against.

**Comment by Shabirmean** [comment]
> How are you running these tests locally / with what variable values? I see
> 
> ```
> Tested values are the variables and the planned_values in the terraform plan
> ```
> 
> But where are you getting these? Do you have test cases?

Currently the tested values are basically the variables set for the module in each test run ([ip module](https://github.com/GoogleCloudPlatform/anthos-samples/pull/26/files#diff-66762536829fb84585036c8874822dbc472c349c91074aabcafbf720e4fbc0a0R46-R47), [init module](https://github.com/GoogleCloudPlatform/anthos-samples/pull/26/files#diff-ce073efb58975c38e91a4a2d9a30455431397eadfa80e87e829e1b2ffb4ef473R52-R64), [vm module](https://github.com/GoogleCloudPlatform/anthos-samples/pull/26/files#diff-3b5f3e89af61cb2a63a1e1a17c0f2643ef182fc5d9ea5de9d6fb1e74a185a342R50-R53)). 

What is tested is:
- if the output of `tf plan` actually has these variables defined _(to verify that other modules using this module won't break)_ 
- if the variables that was passed in is the ones being set
- for list type if there are that many resources created
- whether the output of the plan has the expected outputs defined in the outputs file
- if there are defaults then ensure the default is set when variable is not passed in _(I still have to add tests for this)_

My understanding is that invalid data types and values would be caught by terraform. As in if a value was not provided then terraform would fail. So there isn't a clear way to test that. These unit tests validate the `plan`.

However, in the integration tests _(which will be sort of an extended unit test)_ I would check things like if the passed in _region_, _zone_ is valid etc.

**PS:** I have a separate [branch for integration tests](https://github.com/GoogleCloudPlatform/anthos-samples/tree/modules-tests/anthos-bm-gcp-terraform/test/module):

**Comment by askmeegs** [comment]
Thanks for these responses Shabir, I was able to get the tests working locally 

```
PASS
ok  	github.com/GoogleCloudPlatform/anthos-samples/anthos-bm-gcp-terraform/unit	11.126s
```

Do you want to hold on merging this until you write the `main.tf` tests? Or will that be a separate PR?

**Comment by Shabirmean** [comment]
> Thanks for these responses Shabir, I was able to get the tests working locally
> 
> ```
> PASS
> ok  	github.com/GoogleCloudPlatform/anthos-samples/anthos-bm-gcp-terraform/unit	11.126s
> ```
> 
> Do you want to hold on merging this until you write the `main.tf` tests? Or will that be a separate PR?

Thank you for reviewing it early @askmeegs. 🙌🏼  
Yes, I'm holding it off until the tests are complete. I have been setting up the GH actions and also waiting until I meet with @puneith later today to finalize this strategy

**Comment by Shabirmean** [comment]
@askmeegs This PR is now complete with all tests and is ready for final review

**Review by askmeegs — APPROVED** [review]

File: anthos-bm-gcp-terraform/test/unit/module_vm_test.go [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+	}
+	// verify how many child modules of type google_compute_instance_from_template were present
+	assert.Len(
+		goTester,
+		expectedVMNames,
+		numberOfComputeInstanceModules,
+		"Resource count for module type google_compute_instance_from_template does not match in plan",
+	)
+
+	validateOutputs(goTester, vmInstancePlan.PlannedValues.Outputs)
```
Comment (puneith) [inline anthos-bm-gcp-terraform/test/unit/module_vm_test.go:201]: Generic comment: Will these tests (e.g. validate compute instance sub module) work when run on say other cloud provider or on-prem/disconnected?

Comment (Shabirmean) [reply anthos-bm-gcp-terraform/test/unit/module_vm_test.go:201]: Yes - as long as we can keep the same name for `input & output` variables for the module.

**Review by puneith — APPROVED** [review]
LGTM

**Pull request merged**